Docs 9444 value resolvers (#82)
mulesoft/docs-mule-sdk: modules/ROOT/pages/rest-sdk/valueresolver.adoc

= Define a Value Resolver
ifndef::env-site,env-github[]
include::../_attributes.adoc[]
endif::[]

> // what is a value resolver?

Reviewer: This feature is designed to provide the possible values of an operation’s parameter. 

Value resolvers will allow the connector to make a request to an endpoint and get the possible values from there, providing capabilities to extract the actual values from the server response.

Author: ```suggestion
Value resolvers provide the possible values of an operation’s parameter.

Value resolvers also enable the connector to make a request to an endpoint and get the possible values from there, providing capabilities to extract the actual values from the server response.
```

The user sees *displayName* in the dropdown for the
possible channel values, and *value* is sent to the server.

You can then add this value resolver to the channel
parameter in the operation as shown:

[source,yaml]
----
endpoints:
 /chat.postMessage:
   operations:
     post:
       expects:
         queryParameter:
           channel:
             valueResolver:
               id: channels
----

== Example: Value Resolver with Parameters

This example contains the previous parameter resolver and adds
some intelligence. This value resolver uses the operation’s
parameters in order to query for the appropriate type of channels.

[source,yaml]
----
valueResolvers:
 channels:
   path: /conversations.list
   method: GET
   parameters:
      channelType:
        type: string
   binding:
     queryParams:
       channelType:
         value:
           expression: “#[if(parameters.channelType != null) parameters.channelType else ‘all’]”
   items:
      extraction:
         expression: “#[payload.channels]”
      displayName:
         expression: “#[item.name]”
      value:
         expression: “[item.id]”
----

> This value resolver resembles the previous one with the exception of
> the addition of the `binding` node. +

Author: ```suggestion
This value resolver resembles the previous one with the exception of the addition of the `binding` node. +
```

> You can use this node to bind parameter values from the operation’s where
> this value resolver is used to the request that is made to
> request the possible values.

Author: ```suggestion
You can use this node to bind the operation's parameter values where the value resolver is used to the request that is made to request the possible values.
```

Author: @giordanolucas this is such a complicated way of articulating the concept of binding two different things... Are there references I can use from the example to clarify this? I do get what you're saying, but I think it would be unclear to the user.

Author: resolving the comment under this, so pasting what you said here:

"This is a way of "forwarding" parameters set in the operation to the value resolver, so it can use those values to make a request and get the possible values the parameter is allowed to have."

Reviewer: Hi @sarathecoubian! Would it be clearer like this?

> This is a way of forwarding parameters set in the operation to the value resolver. These bound parameters can be later used to build the Value Resolver request.

Author: Yes! Beautiful! Thank you!!!

Author: ```suggestion
This is a way of forwarding parameters set in the operation to the value resolver. These bound parameters can be later used to build the value resolver request.
```

> // I don't really understand what this sentence is trying to convey ^ Is operations supposed to be possessive?

Reviewer: This is a way of "forwarding" parameters set in the operation to the value resolver, so it can use those values to make a request and get the possible values the parameter is allowed to have.

Author: ```suggestion
```

In this example, the `channelType` query parameter is bound to the
value provided in the operation `uriChannelType` URI parameter.

[source,yaml]
----
/chat.postMessage/{uriChannelType}:
   operations:
     post:
       expects:
         queryParameter:
           channel:
             valueResolver:
               id: channels
               arguments:
                  channelType:
                    Value: uriParameter.uriChannelType
----

> Setting the value resolver to the operation is as simple as in the
> previous example, only that in this case it is validated that the
> binded parameters exist in the operation before generating the connector
> code.

Author: ```suggestion
Setting the value resolver to the operation is as simple as in the previous example, only that in this case it is validated that the bound parameters exist in the operation before generating the connector code.
```

> //"bound" parameters?

Reviewer: Yep

> //"bound" parameters?

Author: ```suggestion
```
